sdks/java/io/csv/src/main/java/org/apache/beam/sdk/io/csv/CsvIO.java

```diff
@@ -0,0 +1,553 @@
+      Optional<PrintWriter> printWriter = Optional.ofNullable(writer);
+      checkState(printWriter.isPresent());
+      return printWriter.get();
+    }
+
+    /**
+     * Not to be confused with the CSV header, it is content written to the top of every sharded
+     * file prior to the header. See {@link Write#withPreamble(String)}.
+     */
+    @Nullable
+    abstract String getPreamble();
+
+    /**
+     * The column names of the CSV file written at the top line of each shard after the preamble, if
+     * available.
+     */
+    abstract String getHeader();
+
+    /** A {@link SerializableFunction} for converting to a CSV formatted string. */
+    abstract SerializableFunction<T, String> getFormatFunction();
+
+    @AutoValue.Builder
+    abstract static class Builder<T> {
+
+      /**
+       * Not to be confused with the CSV header, it is content written to the top of every sharded
+       * file prior to the header. See {@link Write#withPreamble(String)}.
+       */
+      abstract Builder<T> setPreamble(String value);
+
+      /**
+       * The column names of the CSV file written at the top line of each shard after the preamble,
+       * if available.
+       */
+      abstract Builder<T> setHeader(String value);
+
+      /** A {@link SerializableFunction} for converting a to a CSV formatted string. */
+      abstract Builder<T> setFormatFunction(SerializableFunction<T, String> value);
+
+      abstract Sink<T> build();
+    }
+  }
+
+  /** {@link PTransform} for writing CSV files. */
+  @AutoValue
+  public abstract static class Write<T> extends PTransform<PCollection<T>, PDone> {
+
+    /** Specifies a common prefix for all generated files. */
+    public Write<T> to(String filenamePrefix) {
+      Path path = Paths.get(filenamePrefix);
+      String directory = "";
+      String name = "";
+      if (path.getParent() != null) {
+        directory = path.getParent().toString();
+      }
+      if (path.getFileName() != null) {
+        name = path.getFileName().toString();
+      }
+      return toBuilder()
+          .setFileWrite(
+              FileIO.<Row>write()
+                  .to(directory)
+                  .withPrefix(name)
+                  .withSuffix(DEFAULT_FILENAME_SUFFIX))
+          .build();
+    }
+
+    /**
+     * Overrides the default {@link CSVFormat#DEFAULT} with format of the destination CSV file data.
+     */
+    public Write<T> withCSVFormat(CSVFormat format) {
+      return toBuilder().setCSVFormat(format).build();
+    }
+
+    /**
+     * Controls the subset and order of included fields when writing records. Defaults to {@link
+     * Schema#getFieldNames()} of {@link Schema#sorted()} from {@link PCollection} input. {@link
+     * CsvIO.Write} will only validate whether {@link Field#getType()} of the {@link Schema}'s
+     * {@link Field#getName()}, included in schemaFields, matches any {@link #VALID_FIELD_TYPE_SET}.
+     */
+    public Write<T> withSchemaFields(List<String> schemaFields) {
```

Comment:
I think we should drop this field and instead rely on CSVFormat.withHeader and point people to that. This way we don't have a point of confusion where the CSVFormat specifies a header and the user specifies withSchemaFields.

We can explicitly tell people that we will filter down to the set of fields in CSVFormat.withHeader if specified.